src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+          case FunctionType::Substring: return _evaluate_substring(expression.arguments);
+          case FunctionType::Concatenate: return _evaluate_concatenate(expression.arguments);
+        }
+      } else {
+        Fail("Function can only be evaluated to a string");
+      }
+      // clang-format on
+  }
+}
+
+template <>
+std::shared_ptr<ExpressionResult<std::string>> ExpressionEvaluator::_evaluate_extract_expression<std::string>(
+    const ExtractExpression& extract_expression) {
+  const auto from_result = evaluate_expression_to_result<std::string>(*extract_expression.from());
+
+  switch (extract_expression.datetime_component) {
+    case DatetimeComponent::Year:
+      return _evaluate_extract_substr<0, 4>(*from_result);
+    case DatetimeComponent::Month:
+      return _evaluate_extract_substr<5, 2>(*from_result);
+    case DatetimeComponent::Day:
+      return _evaluate_extract_substr<8, 2>(*from_result);
+
+    case DatetimeComponent::Hour:
+    case DatetimeComponent::Minute:
+    case DatetimeComponent::Second:
+      Fail("Hour, Minute and Second not available in String Datetimes");
+  }
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_extract_expression(
+    const ExtractExpression& extract_expression) {
+  Fail("Only Strings (YYYY-MM-DD) supported for Dates right now");
+}
+
+template <size_t offset, size_t count>
+std::shared_ptr<ExpressionResult<std::string>> ExpressionEvaluator::_evaluate_extract_substr(
+    const ExpressionResult<std::string>& from_result) {
+  std::shared_ptr<ExpressionResult<std::string>> result;
+
+  std::vector<std::string> values(from_result.size());
+
+  from_result.as_view([&](const auto& from_view) {
+    for (auto chunk_offset = ChunkOffset{0}; chunk_offset < from_view.size(); ++chunk_offset) {
+      if (!from_view.null(chunk_offset)) {
+        DebugAssert(from_view.value(chunk_offset).size() == 10u,
+                    "Invalid DatetimeString '"s + from_view.value(chunk_offset) + "'");
+        values[chunk_offset] = from_view.value(chunk_offset).substr(offset, count);
+      }
+    }
+  });
+
+  return std::make_shared<ExpressionResult<std::string>>(std::move(values), from_result.nulls);
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_negate_expression(
+    const NegateExpression& negate_expression) {
+  std::vector<R> values;
+  std::vector<bool> nulls;
+
+  _resolve_to_expression_result(*negate_expression.argument(), [&](const auto& argument_result) {
+    using ArgumentType = typename std::decay_t<decltype(argument_result)>::Type;
+
+    // clang-format off
+    if constexpr (!std::is_same_v<ArgumentType, std::string> && std::is_same_v<R, ArgumentType>) {
+      values.resize(argument_result.size());
+      for (auto chunk_offset = ChunkOffset{0}; chunk_offset < argument_result.size(); ++chunk_offset) {
+        values[chunk_offset] = -argument_result.values[chunk_offset];
```

Comment:
First, I think that the `-` is easy to miss here. For clarity, maybe consider writing `-1 * `. 
Second, until now, I thought that negate was responsible for booleans. How does `-x` make sense for booleans?

Reply:
Changed that to 

```

        // NOTE: Actual negation happens in this line
        values[chunk_offset] = -argument_result.values[chunk_offset];
```

The NegateExpression covers ONLY negation of numbers, not "boolean-NOT" Should we rename it to UnaryMinusExpression? 

I will check that NOT (a > 5 AND b< 3) works, right now I'm not sure which path it takes.

Reply:
it takes the 
1. Compute the complex predicate into a 1-or-0-column
2. It runs a TableScan for != 0 on this column.
path.

We will replace this with a direct evaluation to a PosList in a later PR